worker-message/src/worker.rs

```diff
@@ -0,0 +1,325 @@
+use crate::message::{ExecuteOutput, Request, Response};
+use crate::{
+    CoordinatorMessage, ExecuteCommandRequest, ExecuteCommandResponse, JobReport, ReadFileRequest,
+    ReadFileResponse, ResponseResult, StreamCommandRequest, StreamCommandResponse, WorkerMessage,
+    WriteFileRequest, WriteFileResponse,
+};
+use std::path::{Path, PathBuf};
+use std::process::{Output, Stdio};
+
+use anyhow::{Error, Result};
+use std::time::SystemTime;
+use tempdir::TempDir;
+use tokio::fs;
+use tokio::io::{AsyncBufReadExt, AsyncReadExt, AsyncWriteExt, BufReader};
+use tokio::process::{Child, Command};
+use tokio::sync::mpsc;
+use tokio::task::{JoinHandle, JoinSet};
+use tracing::info;
+
+pub async fn listen() {
+    let mut tasks = JoinSet::new();
+    let (tx, mut coordinator_rx) = mpsc::channel(8);
+    tasks.spawn(async move {
+        let mut buffer = Vec::new();
+        let mut input = tokio::io::stdin();
+        loop {
+            let payload_len = input
+                .read_u64()
+                .await
+                .expect("Failed to read standard input");
+            buffer.resize(payload_len as usize, 0);
+            input
+                .read_exact(&mut buffer)
+                .await
+                .expect("Failed to read standard input");
+            let coordinator_msg: CoordinatorMessage =
+                bincode::deserialize(&buffer).expect("Failed to deserialize CoordinatorMessage");
+            tx.send(coordinator_msg).await.ok(/*Only fail due to process exiting*/);
+        }
+    });
+    let (coordinator_tx, mut rx) = mpsc::channel(8);
+    tasks.spawn(async move {
+        let mut output = tokio::io::stdout();
+        loop {
+            let worker_msg = rx.recv().await.expect("Failed to write to standard output");
+            info!("{:#?}", worker_msg);
+            let encoded =
+                bincode::serialize(&worker_msg).expect("Failed to serialize WorkerMessage");
+            output
+                .write_u64(encoded.len() as u64)
+                .await
+                .expect("Failed to write to standard output");
+            output
+                .write_all(&encoded)
+                .await
+                .expect("Failed to write to standard output");
+            output
+                .flush()
+                .await
+                .expect("Failed to flush standard output");
+        }
+    });
+
+    tasks.spawn(async move {
+        let time_in_nanos = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap() // now can't be earlier than UNIX_EPOCH.
+            .as_nanos();
+        let project_dir = TempDir::new(&format!("playground-{time_in_nanos}")).expect("Failed to create temporary project directory");
```

Comment:
This is for running tests concurrently.
Otherwise all tests will write to the same files.

Reply:
But each test should be creating a separate Docker container and thus be completely isolated from each other.

Reply:
Yes. But creating Docker images is slow when you're in dev cycle.

Maybe we can use a envionmental flag to toggle this?

Reply:
Yes, I think that there should be some kind of way to switch because of the exact slow cycle time you described. With a large amount of handwaving, I could see some interface like

```rust
Coordinator::new() // uses the expected Docker commands
Coordinator::with_custom_thing(|| {
    // Create and return a `std::process::Command` that when started will provide the stdin/out/err pipes
})
```

Reply:
Well, I see that this is in the worker. In that case, I'd expect that the worker could just gain an argument for which directory to operate in. Then the test code would use `TempDir` to create separate locations and the production code would be unaware.

Reply:
Good idea.